qutip/tests/test_mcsolve.py

```diff
@@ -30,561 +30,283 @@
-    diff = np.mean(abs(actual_answer - expt) / actual_answer)
-    assert_equal(diff < error, True)
-
-
-def test_MCNoCollFuncStates():
-    "Monte-carlo: Constant H (func format) with no collapse ops (states)"
-    error = 1e-8
-    N = 10  # number of basis states to consider
-    a = destroy(N)
-    H = [a.dag() * a, [a.dag() * a, const_H1_coeff]]
-    psi0 = basis(N, 9)  # initial state
-    c_op_list = []
-    tlist = np.linspace(0, 10, 100)
-    mcdata = mcsolve(H, psi0, tlist, c_op_list, [], ntraj=ntraj)
-    states = mcdata.states
-    expt = expect(a.dag() * a, states)
-    actual_answer = 9.0 * np.ones(len(tlist))
-    diff = np.mean(abs(actual_answer - expt) / actual_answer)
-    assert_equal(diff < error, True)
-
-
-def test_MCCollapseTimesOperators():
-    "Monte-carlo: Check for stored collapse operators and times"
-    N = 10
-    kappa = 5.0
+import qutip
+
+
+def _return_constant(constant, *args, **kwargs):
+    return constant
+
+
+def _return_decay(constant, rate, t, *args, **kwargs):
+    return constant * np.exp(-rate * t)
+
+
+@pytest.mark.usefixtures("in_temporary_directory")
+class StatesAndExpectOutputCase:
+    """
+    Mixin class to test the states and expectation values from ``mcsolve``.
+    """
+    size = 10
+    h = qutip.num(size)
+    state = qutip.basis(size, size-1)
+    times = np.linspace(0, 1, 101)
+    e_ops = [qutip.num(size)]
+    ntraj = 750
+
+    def _assert_states(self, result, expected, tol):
+        assert hasattr(result, 'states')
+        assert len(result.states) == len(self.times)
+        for test_operator, expected_part in zip(self.e_ops, expected):
+            test = qutip.expect(test_operator, result.states)
+            np.testing.assert_allclose(test, expected_part, rtol=tol)
+
+    def _assert_expect(self, result, expected, tol):
+        assert hasattr(result, 'expect')
+        assert len(result.expect) == len(self.e_ops)
+        for test, expected_part in zip(result.expect, expected):
+            np.testing.assert_allclose(test, expected_part, rtol=tol)
+
+    def test_states_and_expect(self, hamiltonian, c_ops, expected, tol):
+        options = qutip.Options(average_states=True, store_states=True)
+        result = qutip.mcsolve(hamiltonian, self.state, self.times,
+                               c_ops=c_ops, e_ops=self.e_ops, ntraj=self.ntraj,
+                               options=options)
+        self._assert_expect(result, expected, tol)
+        self._assert_states(result, expected, tol)
+
+
+class TestNoCollapse(StatesAndExpectOutputCase):
+    """
+    Test that `mcsolve` correctly solves the system when there is a constant
+    Hamiltonian and no collapses.
+    """
+    def pytest_generate_tests(self, metafunc):
+        tol = 1e-8
+        expect = (qutip.expect(self.e_ops[0], self.state)
+                  * np.ones_like(self.times))
+        # Use partial to allow pickling.
+        zero_function = functools.partial(_return_constant, 0)
```

Comment:
I would prefer you use `args` instead of partial. For now there is no real advantage, but when the object version arrive, it will be possible to change `args` without recompiling string. It is already possible with the `rhs_reuse` options, but not used much.

Reply:
Oh yeah, I actually completely forgot about the `args` dictionary when I wrote that. That's a better solution regardless of what comes next.

Reply:
Fixed in #1250.